src/scripts/install_machine.py

```diff
@@ -0,0 +1,329 @@
+
+    with prepare_virt_install_params(args) as params:
+        try:
+            subprocess.check_output(params)
+        except subprocess.CalledProcessError as e:
+            # If virt-install returned non-zero return code, redefine
+            # the VM so that we get back the metadata which enable the 'Install'
+            # button, so that the user can re-attempt installation
+            with tempfile.NamedTemporaryFile() as file:
+                logging.debug("virt-install failed, redefining to renable the 'Install' button")
+                file.write(prevXML)
+                file.flush()
+                virsh(args['connectionName'], "define", file.name)
+            raise e
+
+        assert_vm_exists(args['connectionName'], args['vmName'])
+
+        xml = virsh(args['connectionName'], "dumpxml", "--inactive", args['vmName'])
+
+        inject_metadata(xml.decode())
+
+
+def inject_metadata(xml):
+    # ET.fromstring() already wants UTF-8 encoded bytes
+    root = ET.fromstring(xml)
+    metadata = root.find('metadata')
+    cockpit_machines_metadata = metadata.find('cockpit_machines')
+    if cockpit_machines_metadata:
+        metadata.remove(cockpit_machines_metadata)
+
+    has_install_phase = "true"
+    # VM does not have a pending install phase (visible Install button in the UI) if:
+    # - The script is called from the 'Install' button
+    # - The script is called from the 'Create' dialog and the VM is started (installer will run)
+    # - The script is called from the 'Import' dialog
+    if args['type'] == 'install' or args['startVm'] or args['sourceType'] == 'disk_image':
+        has_install_phase = "false"
+
+    # Register used namespaces
+    ET.register_namespace("cockpit_machines", "https://github.com/cockpit-project/cockpit-machines")
+    ET.register_namespace("libosinfo", "http://libosinfo.org/xmlns/libvirt/domain/1.0")
+
+    METADATA = f'''
+<cockpit_machines:data> \
+  <cockpit_machines:has_install_phase>{has_install_phase}</cockpit_machines:has_install_phase> \
+  <cockpit_machines:install_source_type>{args['sourceType']}</cockpit_machines:install_source_type> \
+  <cockpit_machines:install_source>{args['source']}</cockpit_machines:install_source> \
+  <cockpit_machines:os_variant>{args['os']}</cockpit_machines:os_variant> \
+</cockpit_machines:data>
+'''
+
+    cockpit_machines_metadata_new = ET.fromstring(METADATA)
+    metadata.append(cockpit_machines_metadata_new)
+
+    updated_xml = ET.tostring(root)
+
+    with tempfile.NamedTemporaryFile() as file:
+        file.write(updated_xml)
+        file.flush()
+        virsh(args['connectionName'], "define", file.name)
+
+
+logging.basicConfig(level=logging.ERROR, format='%(message)s')
+logging.debug(sys.argv[1])
+
+args = json.loads(sys.argv[1], strict=False)
+
+logging.debug(args)
+
+if args['type'] == 'create':
+    create_vm(args)
+elif args['type'] == 'install':
+    install_vm(args)
```

Comment:
Now the script silently does nothing for unknown `type`s, instead of installing. That's slightly better, but I think it'd be even better to do something like `raise NotImplementedError("unknown type " + args['type'])`

Reply:
Fixed